crates/core/tedge_actors/src/builders.rs

```diff
@@ -55,6 +56,16 @@ pub trait Builder<T>: Sized {
     }
 }
 
+pub trait ActorBuilder {
+    type Error: std::error::Error;
+
+    /// Prepare any initial state for the actor before the actor is spawned.
+    /// For e.g, creating initial state files/directories etc
+    /// The actions performed in this function must be idempotent as it may be called multiple times,
+    /// although typically it is only called once before the actor is spawned.
+    fn init(config_dir: &Path) -> Result<(), Self::Error>;
```

Comment:
The `init` API in placed here in the `ActorBuilder` instead of the `Actor` itself as the existing init logic in our plugins are performed during the installation of the plugin when we don't have all the configurations(tedge.config) to fully initialize the actor itself. Doing this `init` from the builder actually limits its capabilities to do only perform static operations like creating files/directories only. It can't perform any advanced init logic like interacting with MQTT, HTTP etc as a fully configured tedge config would be required for an actor to perform all such things.

But, this raises one question: What do we gain by doing such a "partial initialization" as part of the installation itself? Why not expect the actor to prepare its initial state as part of its `run` logic itself, before it starts processing requests?

Reply:
> But, this raises one question: What do we gain by doing such a "partial initialization" as part of the installation itself? Why not expect the actor to prepare its initial state as part of its run logic itself, before it starts processing requests?

Indeed, a crucial question! 

* The `--init` options of the mappers and agent are currently not idempotent - hence buggy, because creating an MQTT sessions and subscriptions, also consume* messages. This has been introduced to avoid race conditions on start of the mappers and agent (with messages sent on topics for which no subscriptions has

Reply:
Yes I agree with suggestions here.

As we slowly abstract our way away from the filesystem then this becomes less relevant. For example if we have a MQTT interface for components to register their "operations" then there is no longer need to think about the file permissions on the folder where they are stored as it is managed by the responsible component.

Though there might be an interesting aspect when it comes to pre-configuring devices, then a runtime interface would make things more difficult. In that case a `tedge init` command might be good, but it would not make it mandatory to use.